drivers/SmartThings/matter-switch/src/init.lua

```diff
@@ -955,12 +977,47 @@ local function y_attr_handler(driver, device, ib, response)
   else
     local y = ib.data.value
     local h, s, _ = color_utils.safe_xy_to_hsv(x, y)
-    device:emit_event_for_endpoint(ib.endpoint_id, capabilities.colorControl.hue(h))
-    device:emit_event_for_endpoint(ib.endpoint_id, capabilities.colorControl.saturation(s))
+    local current_color_mode = get_field_for_endpoint(device, COLOR_MODE, ib.endpoint_id)
+    -- don't send capability events if the color of the device is being determined by CurrentHue and CurrentSaturation
+    -- (ColorMode 0) or by ColorTemperatureMireds (ColorMode 2), store the values for now in case the ColorMode was changed.
+    set_field_for_endpoint(device, CURRENT_HUE, ib.endpoint_id, h, {persist = true})
+    set_field_for_endpoint(device, CURRENT_SAT, ib.endpoint_id, s, {persist = true})
+    if current_color_mode ~= 0 and current_color_mode ~= 2 then
+      device:emit_event_for_endpoint(ib.endpoint_id, capabilities.colorControl.hue(h))
+      device:emit_event_for_endpoint(ib.endpoint_id, capabilities.colorControl.saturation(s))
+    end
     device:set_field(RECEIVED_X, nil)
   end
 end
 
+local function color_mode_attr_handler(driver, device, ib, response)
+  if ib.data.value ~= nil then
+    local previous_color_mode = get_field_for_endpoint(device, COLOR_MODE, ib.endpoint_id)
+    local hue = get_field_for_endpoint(device, CURRENT_HUE, ib.endpoint_id)
+    local sat = get_field_for_endpoint(device, CURRENT_SAT, ib.endpoint_id)
+    if ib.data.value ~= previous_color_mode then
+      if ib.data.value == 0 then
+        if hue ~= nil then
+          device:emit_event_for_endpoint(ib.endpoint_id, capabilities.colorControl.hue(hue))
+          set_field_for_endpoint(device, CURRENT_HUE, ib.endpoint_id, nil, {persist = true})
+        end
+        if sat ~= nil then
+          device:emit_event_for_endpoint(ib.endpoint_id, capabilities.colorControl.saturation(sat))
+          set_field_for_endpoint(device, CURRENT_SAT, ib.endpoint_id, nil, {persist = true})
+        end
+      elseif ib.data.value == 1 then
+        if hue ~= nil and sat ~= nil then
+          device:emit_event_for_endpoint(ib.endpoint_id, capabilities.colorControl.hue(hue))
+          device:emit_event_for_endpoint(ib.endpoint_id, capabilities.colorControl.saturation(sat))
+          set_field_for_endpoint(device, CURRENT_HUE, ib.endpoint_id, nil, {persist = true})
+          set_field_for_endpoint(device, CURRENT_SAT, ib.endpoint_id, nil, {persist = true})
+        end
+      end
```

Comment:
I think we may want a handler for 2 (`ColorTemperatureMireds`) that clears out the saved hue/sat so that later when it switches back to 0 or 1 we don't emit the old hue/sat values